gap/SnAnUnknownDegree.gi

```diff
@@ -0,0 +1,710 @@
+# FIXME: Move me into GAP
+# Helper function to compute all primes up to a given integer via a prime sieve
+BindGlobal("AllPrimesUpTo",
+function(n)
+    local i, j, sieve, result;
+    if n <= 1000 then
+      return Primes{[1..PositionSorted(Primes, n+1)-1]};
+    fi;
+    sieve := BlistList([1..n], [1..n]);
+    sieve[1] := false;
+    for i in [2..Int(n/2)] do
+        sieve[i*2] := false;
+    od;
+    i := 3;
+    while i * i <= n do
+        if sieve[i] then
+            j := 3*i;
+            while j <= n do
+                sieve[j] := false;
+                j := j + 2*i;
+            od;
+        fi;
+        i := i + 2;
+    od;
+    return ListBlist([1..n], sieve);
+end);
+
+
+# ri : recog info record with group G
+# eps : real number, the error bound
+# N : integer, upper bound for the degree of G
+#
+# The following algorithm constructs a set of possible 3-cycles. It is based
+# on the simple observation that the product of two involutions t1, t2, which
+# only move one common point, squares to a 3-cycle.
+#
+# TODO: take care of duplicate candidates?
+# Creates and returns a function, here called oneThreeCycleCandidate. The
+# function oneThreeCycleCandidate returns one of the following:
+# - a three cycle candidate, i.e. an element of G
+# - TemporaryFailure, if we exhausted all attempts
+# - NeverApplicable, if we found out that G can't be an Sn or An
+BindGlobal("ThreeCycleCandidatesIterator",
+    function(ri, eps, N)
+    local
+        # involution
+        t,
+        # integers, controlling the number of iterations
+        M, B, T, C, logInt2N,
+        # iteration over all primes up to N
+        allPrimes, i, p,
+        # counters
+        nrInvolutions, nrTriedConjugates, nrThreeCycleCandidates,
+        # helper functions
+        tryThreeCycleCandidate, oneThreeCycleCandidate;
+    # Step 1: Initialization
+    # The current involution t_i
+    t := fail;
+
+    # Constants
+    # TODO: better iteration over primes
+    M := 1;
+    allPrimes := AllPrimesUpTo(N);
+    for i in [2 .. Length(allPrimes)] do
+        p := allPrimes[i];
+        M := M * p ^ LogInt(N, p);
+    od;
+    # FIXME: Probably B can be chosen smaller
```

Comment:
Where does this opinion come from / how is it justified?

Reply:
@FriedrichRober  mentioned sometime that @aniemeyer may know a trick to choose a smaller B. We still have to set up a proper mail to ask her that though. :)